### ISSUE TITLE
Crash when built with GCC and OpenMP on OS X

### ISSUE THREAD

**Issue by mikepb** [explosion/spaCy#266]
I was testing spaCy's parallel pipe feature on an unparsed XML dump of Wikipedia and Python crashed with `EXC_BAD_ACCESS`. spaCy was compiled with OpenMP support using GCC 5.3.0.

Steps to reproduce:

``` sh
brew install gcc  --without-multilib
brew install python
export CC=gcc-5
export CXX=g++-5
```

Edit `setup.py` lines 88-90 to enable OpenMP:

``` py
# if not sys.platform.startswith('darwin'):
compile_options['other'].append('-fopenmp')
link_options['other'].append('-fopenmp')
```

Follow the 'Compile from source' [instructions](https://spacy.io/#install), ensuring that `CC` and `CXX` have been set correctly. The build log should show `gcc-5` and `g++-5` being used instead of `clang`.

Run this script:

``` py
import bz2
import itertools
import operator
import spacy.en

f = bz2.BZ2File('enwiki-20160204-pages-articles.xml.bz2', 'rb')
texts = itertools.imap(operator.methodcaller('decode', 'utf-8'), f)
nlp = spacy.en.English(entity=False)

print('Piping texts...')
docs = nlp.pipe(texts, n_threads=-1)

print('Iterating over docs...')
for i, doc in enumerate(docs):
    assert(doc.is_parsed)
```

System information:

```
$ pip2 show spacy
---
Metadata-Version: 2.0
Name: spacy
Version: 0.100.5
Summary: Industrial-strength NLP
Home-page: https://spacy.io
Author: Matthew Honnibal
Author-email: matt@spacy.io
License: MIT
Location: /usr/local/lib/python2.7/site-packages
Requires: cymem, preshed, numpy, thinc, six, text-unidecode, murmurhash, cloudpickle, plac, ujson, sputnik

$ uname -a
Darwin Orion.local 15.3.0 Darwin Kernel Version 15.3.0: Thu Dec 10 18:40:58 PST 2015; root:xnu-3248.30.4~1/RELEASE_X86_64 x86_64

Orion:spaCy mikepb$ python2
Python 2.7.11 (default, Jan 15 2016, 00:22:11) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> platform.version()
'Darwin Kernel Version 15.3.0: Thu Dec 10 18:40:58 PST 2015; root:xnu-3248.30.4~1/RELEASE_X86_64'
>>> platform.mac_ver()
('10.11.3', ('', '', ''), 'x86_64')
>>> platform.python_build()
('default', 'Jan 15 2016 00:22:11')
>>> platform.python_compiler()
'GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)'
>>> platform.python_implementation()
'CPython'
>>> platform.python_version()
'2.7.11'

Orion:spaCy mikepb$ env
TERM_PROGRAM=Apple_Terminal
TERM=xterm-256color
SHELL=/bin/bash
TMPDIR=/var/folders/yz/1_3j3f_x7x938w7_b41vwrpm0000gn/T/
Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.g31CSGBfIf/Render
TERM_PROGRAM_VERSION=361.1
TERM_SESSION_ID=3FE42C09-E758-4070-8012-F3C6FD15EC94
USER=mikepb
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.eytmZ6Ieij/Listeners
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x0
PATH=./node_modules/.bin:/usr/local/opt/node/bin:/Users/mikepb/Developer/gopath/bin:/Library/TeX/Distributions/Programs/texbin:/usr/local/sbin:/usr/local/bin:/usr/local/cuda/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin
PWD=/Users/mikepb/Developer/spaCy
LANG=en_US.UTF-8
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
HOME=/Users/mikepb
SHLVL=1
LOGNAME=mikepb
_=/usr/bin/env
OLDPWD=/Users/mikepb/Developer
```

See attached for Python script and logs:
- [crash_test.zip](https://github.com/spacy-io/spaCy/files/136840/crash_test.zip)

**Comment by honnibal** [comment]
Great report — thanks. Looks like the problem is here:

https://github.com/spacy-io/spaCy/blob/master/setup.py#L84
```python
if not sys.platform.startswith('darwin'):
    compile_options['other'].append('-fopenmp')
    link_options['other'].append('-fopenmp')
```

Maybe this should be `sys.platform.lower().startswith('darwin')`? @henningpeters ?

**Comment by mikepb** [comment]
`sys.platform` is already lowercase:

```
>>> import sys
>>> sys.platform
'darwin'
```

**Comment by henningpeters** [comment]
```sys.platform``` is always lowercase, see https://docs.python.org/2/library/sys.html#sys.platform
@mikepb are you sure your gcc has openmp support (should be compiled with ```--without-multilib```)?

**Comment by mikepb** [comment]
Yes, my gcc was complied with `--without-multilib`.

OK, so I figured out that running the script inside the spaCy source directory actually loads the wrong spaCy module. This is a non-issue.

I was looking into enable OpenMP on OS X using GCC. I'll open a separate issue for that.

**Comment by mikepb** [comment]
Reopening as GCC on OS X specific issue for #267

**Comment by mikepb** [comment]
My guess is that this is an upstream issue with cython?

**Comment by henningpeters** [comment]
Thanks for the extensive analysis. Unfortunately, I cannot reproduce the issue (using only a small extract of Wikipedia though: ```enwiki-20160204-pages-articles1.xml-p000000010p000030302.bz2```, full Wikipedia download still going on. Also using ```gcc 5.3.0 --without-multilib```, but in my case additionally I had to shadow some compiler flags: ```-Wshorten-64-to-32``` and ```-arch i386```. This depends on how your Python was built I think. Let's check Cython versions, mine is ```0.23.4```.

**Comment by mikepb** [comment]
```
$ cython --version
Cython version 0.23.4
```

**Comment by mikepb** [comment]
I found that I had to delete the cython-generated files to get things to compile right:

```sh
find spacy \( -name '*.so' -or -name '*.cpp' -or -name '*.pyc' \) -delete
```

**Comment by mikepb** [comment]
Also, I'm using brew-installed Python:

```
$ brew info python
python: stable 2.7.11 (bottled), HEAD
Interpreted, interactive, object-oriented programming language
https://www.python.org
/usr/local/Cellar/python/2.7.11 (6,197 files, 83.4M) *
  Built from source
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/python.rb
==> Dependencies
Build: pkg-config ✔
Required: openssl ✔
Recommended: readline ✔, sqlite ✔, gdbm ✔
Optional: homebrew/dupes/tcl-tk ✘, berkeley-db4 ✘
==> Options
--universal
	Build a universal binary
--with-berkeley-db4
	Build with berkeley-db4 support
--with-poll
	Enable select.poll, which is not fully implemented on OS X (https://bugs.python.org/issue5154)
--with-quicktest
	Run `make quicktest` after the build (for devs; may fail)
--with-tcl-tk
	Use Homebrew's Tk instead of OS X Tk (has optional Cocoa and threads support)
--without-gdbm
	Build without gdbm support
--without-readline
	Build without readline support
--without-sqlite
	Build without sqlite support
--HEAD
	Install HEAD version
==> Caveats
Pip and setuptools have been installed. To update them
  pip install --upgrade pip setuptools

You can install Python packages with
  pip install <package>

They will install into the site-package directory
  /usr/local/lib/python2.7/site-packages

See: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Homebrew-and-Python.md

.app bundles were installed.
Run `brew linkapps python` to symlink these to /Applications.```

**Comment by henningpeters** [comment]
There is a cleanup shortcut: ```python setup.py clean```, this also removes .so and .cpp

**Comment by henningpeters** [comment]
Was your Python compiled with gcc or clang?

**Comment by henningpeters** [comment]
Ah sorry, its in your log: ```[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)]```, so clang.

**Comment by mikepb** [comment]
Good to know about the shortcut!

Python was downloaded as pre-compiled from brew, using LLVM. With LLVM-built R, I've used native extensions built with both GCC and LLVM. As far as I'm aware, they all use C bindings, though. It's nice that the C ABI is well defined just for this.

Could there be an issue with GCC C++ and LLVM Python?

Here's a LLVM run log:

```
(.env) Orion:spaCy mikepb$ lldb python2 crash_test.py 
(lldb) target create "python2"
Current executable set to 'python2' (x86_64).
(lldb) settings set -- target.run-args  "crash_test.py"
(lldb) run
Process 1918 launched: '/Users/mikepb/Developer/spaCy/.env/bin/python2' (x86_64)
Piping texts...
Iterating over docs...
Process 1918 stopped
* thread #1: tid = 0x653c, 0x0000000101fa4cf1 libgomp.1.dylib`gomp_team_start + 993, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7f975fbff240)
    frame #0: 0x0000000101fa4cf1 libgomp.1.dylib`gomp_team_start + 993
libgomp.1.dylib`gomp_team_start:
->  0x101fa4cf1 <+993>:  movl   $0x0, 0x60(%r13)
    0x101fa4cf9 <+1001>: movl   %r14d, %ebx
    0x101fa4cfc <+1004>: movl   %eax, 0x34(%r13)
    0x101fa4d00 <+1008>: movl   0x38(%r15), %eax
(lldb) bt
* thread #1: tid = 0x653c, 0x0000000101fa4cf1 libgomp.1.dylib`gomp_team_start + 993, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7f975fbff240)
  * frame #0: 0x0000000101fa4cf1 libgomp.1.dylib`gomp_team_start + 993
    frame #1: 0x0000000101fa151a libgomp.1.dylib`GOMP_parallel + 58
    frame #2: 0x000000010428d9fe parser.so`::__pyx_gb_5spacy_6syntax_6parser_6Parser_12generator(__pyx_generator=0x0000000103993380, __pyx_sent_value=<unavailable>) + 1134 at parser.cpp:6477
    frame #3: 0x000000010330a4f5 tokenizer.so`(self=0x0000000103993380, value=<unavailable>)(__pyx_CoroutineObject *, PyObject *) + 149 at tokenizer.cpp:12280
    frame #4: 0x000000010008e3db Python`PyEval_EvalFrameEx + 24948
    frame #5: 0x000000010002421b Python`gen_send_ex + 191
    frame #6: 0x00000001000207c7 Python`enum_next + 32
    frame #7: 0x000000010008e3db Python`PyEval_EvalFrameEx + 24948
    frame #8: 0x0000000100088075 Python`PyEval_EvalCodeEx + 1583
    frame #9: 0x0000000100087a40 Python`PyEval_EvalCode + 54
    frame #10: 0x00000001000ac041 Python`run_mod + 53
    frame #11: 0x00000001000ac0e4 Python`PyRun_FileExFlags + 133
    frame #12: 0x00000001000abc33 Python`PyRun_SimpleFileExFlags + 698
    frame #13: 0x00000001000bd531 Python`Py_Main + 3137
    frame #14: 0x00007fff820265ad libdyld.dylib`start + 1
    frame #15: 0x00007fff820265ad libdyld.dylib`start + 1
```

**Comment by honnibal** [comment]
Correct me if I'm wrong, but as I read your logs you're not using GCC --- you're using clang. That's what `gcc` aliases to on OSX.

**Comment by mikepb** [comment]
Looking over the install log, I actually can't tell what compiler is being used. The build command did not output any logs. Here's an excerpt from a new complete build log using the latest master:

```
gcc-5 -fno-strict-aliasing -fno-common -dynamic -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/include/python2.7 -I/Users/mikepb/Developer/spaCy/include -I/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/include/python2.7 -c spacy/syntax/ner.cpp -o build/temp.macosx-10.11-x86_64-2.7/spacy/syntax/ner.o -O3 -Wno-strict-prototypes -Wno-unused-function -fopenmp
```

So good thing `pip` respects `CC=gcc-5 CXX=c++-5` and builds with the right compiler.

Here's the full transcript: [spacy_build_gcc_osx_log.txt](https://github.com/spacy-io/spaCy/files/137209/spacy_build_gcc_osx_log.txt)
